daffodil-io/src/main/scala/org/apache/daffodil/io/InputSource.scala

```diff
@@ -340,6 +390,11 @@ class BucketingInputSource(inputStream: java.io.InputStream, bucketSize: Int = 1
       while (bytesStillToGet > 0) {
         val bytesToGetFromCurrentBucket = Math.min(bucketSize - byteIndex, bytesStillToGet).toInt
 
+        if ((bucketIndex < 0) || (buckets(bucketIndex.toInt) == null)) {
+          _isValid = false
+          throw new BacktrackingException("Attempted to backtrack to position %d, which exceeds maximum backtrack length of %d".format(bucketIndex, maxCacheSizeInBytes))
```

Comment:
This duplicates the backtracking string. Thoughts on having the BacktrackingException just accept a psotion and max backtrack length, and then the exception can build the string in the constructor?